src/backends/plonky2/primitives/ec/bits.rs

```diff
@@ -0,0 +1,200 @@
+}
+
+impl CircuitBuilderBits for CircuitBuilder<GoldilocksField, 2> {
+    fn conditional_zero(&mut self, if_zero: Target, then_zero: Target) {
+        let quot = self.add_virtual_target();
+        self.add_simple_generator(ConditionalZeroGenerator {
+            if_zero,
+            then_zero,
+            quot,
+            _phantom: PhantomData,
+        });
+        let prod = self.mul(if_zero, quot);
+        self.connect(prod, then_zero);
+    }
+
+    fn biguint_bits(&mut self, x: &BigUInt320Target) -> [BoolTarget; 320] {
+        let bits = x.0.map(|t| self.low_bits(t, 32, 32));
+        array::from_fn(|i| bits[i / 32][i % 32])
+    }
+
+    fn field_elements_to_biguint(&mut self, arr: &[Target]) -> BigUInt320Target {
+        assert!(arr.len() <= 5);
+        let mut ans = BigUInt320Target(array::from_fn(|_| self.zero()));
+        let neg_one = self.neg_one();
+        let two_32 = self.constant(GoldilocksField::from_canonical_u64(1 << 32));
+        for (n, &x) in arr.iter().rev().enumerate() {
+            // multiply by the order of the Goldilocks field
+            for i in (0..(2 * n)).rev() {
+                let tmp = self.add(ans.0[i + 1], two_32);
+                ans.0[i + 1] = self.sub(tmp, ans.0[i]);
+                let tmp = self.add(ans.0[i + 2], ans.0[i]);
+                ans.0[i + 2] = self.add(tmp, neg_one);
+            }
+            // add x
+            let [low, high] = self.split_32_bit(x);
+            ans.0[0] = self.add(ans.0[0], low);
+            ans.0[1] = self.add(ans.0[1], high);
+            self.normalize_bigint(&mut ans, 34, 2 * n + 1);
+        }
+        ans
+    }
+
+    fn normalize_bigint(
+        &mut self,
+        x: &mut BigUInt320Target,
+        max_digit_bits: usize,
+        max_num_carries: usize,
+    ) {
+        for i in 0..max_num_carries {
+            let (low, high) = self.split_low_high(x.0[i], 32, max_digit_bits);
+            x.0[i] = low;
+            x.0[i + 1] = self.add(x.0[i + 1], high);
+        }
+    }
+
+    fn split_32_bit(&mut self, x: Target) -> [Target; 2] {
+        let (low, high) = self.split_low_high(x, 32, 64);
+        let max = self.constant(GoldilocksField::from_canonical_i64(0xFFFFFFFF));
+        let high_minus_max = self.sub(high, max);
+        self.conditional_zero(high_minus_max, low);
+        [low, high]
+    }
+
+    fn constant_biguint320(&mut self, n: &BigUint) -> BigUInt320Target {
+        assert!(n.bits() <= 320);
+        let digits = n.to_u32_digits();
+        let targets = array::from_fn(|i| {
+            let d = digits.get(i).copied().unwrap_or(0);
+            self.constant(GoldilocksField::from_canonical_u32(d))
+        });
+        BigUInt320Target(targets)
+    }
+
+    fn add_virtual_biguint320_target(&mut self) -> BigUInt320Target {
+        let targets = self.add_virtual_target_arr();
+        for t in targets {
+            self.range_check(t, 32);
```

Comment:
This `range_check` internally is doing `self.split_le(x, 32)`, so internally it's decomposing the target into bits to do the range check.

Then in `biguint_bits` we're calling `self.low_bits` which again calls `self.split_le(x, 32)`, but we already had the bit decomposition of each limb from the constructor!  So perhaps in `BigUInt320Target` we can store the bit decomposition we get from the range check in the constructor, and avoid duplication of constraints?